demos/preconditioning/demo_heat_rana.py.rst

```diff
@@ -0,0 +1,123 @@
+====================================================
+
+This demo applies a method suggested in:
+
+Rana, Howle, Long, Meek, Milestone "A new block preconditioner for implicit Runge-Kutta methods for parabolic PDE problems," SISC 2021
+
+to our ongoing heat equation demonstration problem on :math:`\Omega = [0,10]
+\times [0,10]`, with boundary :math:`\Gamma`, giving rise to the weak form
+
+.. math::
+
+   (u_t, v) + (\nabla u, \nabla v) = (f, v)
+
+A multi-stage RK method applied to the heat equation gives a
+block-structured system.  The on-diagonal blocks are quite similar to
+what one obtains from a backward Euler discretization of the equation.
+
+With a 2-stage method, we have
+
+.. math::
+   
+   \left[ \begin{array}{cc} A_{11} & A_{12} \\ A_{21} & A_{22} \end{array} \right]
+   \left[ \begin{array}{c} k_1 \\ k_2 \end{array} \right]
+   &= \left[ \begin{array}{c} f_1 \\ f_2 \end{array} \right]
+
+The suggestion in the paper is to approximate the Butcher matrix A with
+with a triangular approximation.  For example, if A = LDU, then one could approximate A with LD or DU.  This gives rise to a block triangular preconditioner
+
+.. math::
+
+  P = \left[ \begin{array}{cc} \tilde{A}_{11} & 0 \\ \tilde_{A}_{21} & A_{22} \end{array} \right]
+
+
+This allows one to leverage an existing methodology for a low order
+method like backward Euler for the diagonal blocks.  Empirical results
+suggest that this method is strongly stage-independent.
+
+Common set-up for the problem::
+
+  from firedrake import *  # noqa: F403
+  from ufl.algorithms.ad import expand_derivatives
+  from irksome import LobattoIIIC, TimeStepper, Dt
+
+  butcher_tableau = LobattoIIIC(3)
+
+  N = 16
+  dt = Constant(10. / N)
+  t = Constant(0.0)
+
+  x0 = 0.0
+  x1 = 10.0
+  y0 = 0.0
+  y1 = 10.0
+
+  msh = RectangleMesh(N, N, x1, y1)
+  V = FunctionSpace(msh, "CG", 1)
+  x, y = SpatialCoordinate(msh)
+
+  S = Constant(2.0)
+  C = Constant(1000.0)
+  B = (x-Constant(x0))*(x-Constant(x1))*(y-Constant(y0))*(y-Constant(y1))/C
+  R = (x * x + y * y) ** 0.5
+
+  uexact = B * atan(t)*(pi / 2.0 - atan(S * (R - t)))
+  rhs = expand_derivatives(diff(uexact, t)) - div(grad(uexact))
+  u = interpolate(uexact, V)
+
+  v = TestFunction(V)
+  F =  inner(Dt(u), v)*dx + inner(grad(u), grad(v))*dx - inner(rhs, v)*dx
+
+  bc = DirichletBC(V, 0, "on_boundary")
+
+Now, we define the solver parameters.  We get at the Rana method
+through an Irksome-provided Python preconditioner.  This method
+inherits from :class:`firedrake.AuxiliaryOperatorPC` (it provides the
+Jacobian of the variational form with the approximate Butcher tableu
+substituted) and so provides the user with an 'aux' preconditioner
+to configure.  Since the Rana technique gives us a block triangular
+matrix, a multiplicative field split exactly apply the preconditioner
+if its diagonal blocks are exactly inverted::
+
+  params = {"mat_type": "aij",
+            "snes_type": "ksponly",
+            "ksp_type": "gmres",
```

Comment:
Well, I think I was trying to use the least expensive method everywhere (not that the extra Krylov vectors matter for `fgmres` when the preconditioner is so heavyweight...), and the MG options don't imply a specifically linear operator.  Perhaps we could re-do the MG parameters to force the PC to be linear and then use `gmres` everywhere.

Reply:
Come to think of it, the MG parameters could stand to be put into a nested form, anyway...

Reply:
Take a look -- I nested the parameters and fixed iteration counts/convergence tests on the multigrid levels so that out gmres has to work.  All the PC demos now use gmres.

Reply:
I like consistency more than I dislike plain gmres, so I think this is fine.

Reply:
We'll not quibble over the letter.